### ISSUE TITLE
Specifying JOIN ON-clause without table names in any order of equation sides

### ISSUE THREAD

**Issue by zlobober** [yandex/ClickHouse#5976]
Consider the following execution: https://pastebin.com/ADkjqE32

Issue here is that I should mention equation sides in ON-clause in specific order: a = c, but not c = a. I believe that the initial intention here was to motivate user specifying table names before column names: if I write t2.c = t1.a, it works perfectly. 

As for me, there are some things that may be done better here:

1. Error message may at least mention which table lacks specified columns.
2. When column tables may be deduced unambiguously, it should be OK to use such columns in any side of the equation (this is the case for the listing above).
3. Anyway, such behavior should be well-documented somewhere around the description of JOIN-section.

This issue is probably related to https://github.com/yandex/ClickHouse/issues/4567

**Comment by 4ertus2** [comment]
```
create table t1 (a UInt32, b String) engine = Memory;
create table t2 (c UInt32, d String) engine = Memory;
select * from t1 join t2 on c == a;

DB::Exception: Missing columns: 'a' while processing query: 'a', required columns: 'a', source columns: 'd' 'c'.
```

**Comment by 4ertus2** [comment]
It would be fixed in #6131
